src/shared/LuaLib.inl

```diff
@@ -0,0 +1,351 @@
+    lua_setfield(L, metatable, "__metatable"); //[metatable = 2] -> t[k] = v; t = [2 = ClassMT], k = "__metatable", v = [3 = 1]; pop [3]
+
+    lua_pushcfunction(L, index); //index = cfunction -> [3] = cfunction
+    lua_setfield(L, metatable, "__index"); //[metatable = 2] -> t[k] = v; t = [2], k = "__index", v = cfunc; pop [3]
+
+    lua_pushcfunction(L, newindex);
+    lua_setfield(L, metatable, "__newindex");
+
+    lua_pushcfunction(L, gc_T);
+    lua_setfield(L, metatable, "__gc");
+
+    lua_pushcfunction(L, tostring_T);
+    lua_setfield(L, metatable, "__tostring");
+
+    ExtraInit<T>(L,metatable); //optionally implemented by individual types
+
+    lua_newtable(L); //for method table -> [3] = this table
+    lua_setmetatable(L, methods); //[methods = 1] -> metatable for [1] is [3]; [3] is popped off, top = [2]
+
+    _regfunctions(L,metatable,methods);
+
+    lua_pop(L, 2); //remove the two items from the stack, [1 = methods] and [2 = metatable]
+}
+
+
+template<typename T>
+int LuaLib<T>::push(lua_State *L, T* obj, bool gc)
+{
+    //for annotations, starting at index 1, but it is a relative number, not always 1
+    if (!obj) { lua_pushnil(L); return lua_gettop(L); }
+    luaL_getmetatable(L, GetTClassName<T>());  // lookup metatable in Lua registry ->[1] = metatable of <ClassName>
+    if (lua_isnil(L, -1)) luaL_error(L, "%s missing metatable", GetTClassName<T>());
+    int mt = lua_gettop(L); //mt = 1
+    T** ptrHold = (T**)lua_newuserdata(L,sizeof(T**)); //->[2] = empty userdata
+    int ud = lua_gettop(L); //ud = 2
+    if(ptrHold != NULL)
+    {
+        *ptrHold = obj;
+        lua_pushvalue(L, mt); // ->[3] = copy of [1]
+        lua_setmetatable(L, -2); //[-2 = 2] -> [2]'s metatable = [3]; pop [3]
+        char name[32];
+        tostring(name,obj);
+        lua_getfield(L,LUA_REGISTRYINDEX,"DO NOT TRASH"); //->[3] = value returned from function
+        if(lua_isnil(L,-1) ) //if [3] hasn't been created yet, then create it
+        {
+            luaL_newmetatable(L,"DO NOT TRASH"); //[4] = the new metatable
+            lua_pop(L,1); //pop [4]
+        }
+        lua_pop(L,1); //pop [3]
+        lua_getfield(L,LUA_REGISTRYINDEX,"DO NOT TRASH"); //->[3] = value returned from function
+        if(gc == false) //if we shouldn't garbage collect it, then put the name in to [3]
+        {
+            lua_pushboolean(L,1);// ->[4] = true
+            lua_setfield(L,-2,name); //represents t[k] = v, [-2 = 3] = t -> v = [4], k = <ClassName>; pop [4]
+        }
+        else
+        {
+            //In case this is an address that has been pushed
+            //to lua before, we need to set it to nil
+            lua_pushnil(L); // ->[4] = nil
+            lua_setfield(L,-2,name); //represents t[k] = v, [-2 = 3] = t -> v = [4], k = <ClassName>; pop [4]
+        }
+
+        if(IsReferenceCounted<T>())
+        {
+			// TODO: Remove
```

Comment:
So what is the story with object life cycle? Do we need that "DO NOT TRASH" shit? I don't remember how it works but if so we might have a problem since this library appears to be based on a really old version that doesn't have my [crash fix](https://github.com/mikke89/RmlUi/commit/23f2e148ec9730e353b520c665c1d1aa089b8cc6) (and even still uses the `Rocket` namespace).

Also when importing code like this it would be nice to have the unmodified version of the imported code (not included in the build) as the first commit, with a note as to which commit in the external project it came from, and only hack it to suit our needs in the second commit.